watershed_workflow/mesh.py

```diff
@@ -1565,3 +1565,85 @@ def create_submesh(m2, shp):
     # create the new mesh
     new_mesh = Mesh2D(new_coords, new_conns, new_labeled_sets, m2.crs, m2.eps, False, True)
     return new_coords_map, new_conn_map, new_mesh
+
+
+def merge_meshes(meshes, transfer_labeled_sets=True):
+    """ Combines multiple 2D meshes into a single mesh. 
+
+    It is assumed that the meshes to be combined have common vertices
+    on the shared edge (no steiner points). Ids of the entities in the
+    labeled set for each mesh to be merged are also updated and assigned 
+    to the combined mesh
+    
+    Parameters
+    ----------
+    meshes : list(mesh.Mesh2D)
+      The list of meshes to be merged
+    transfer_labeled_sets : boolean, optional
+      if true, the labeled sets from individual meshes are updated
+      and assigned to the combined mesh
+
+    Returns
+    -------
+    mesh.Mesh2D
+      combined mesh
+    """
+    # ## identify base mesh
+    # meshes = list(sorted(meshes, key=lambda m : -m.num_cells)) ##FIX ME##
+
+    m2_combined = meshes[0]
+    for mesh in meshes[1:]:
+        m2_combined  = merge_mesh(m2_combined , mesh, transfer_labeled_sets=transfer_labeled_sets)
+
+    return m2_combined 
+
+
+def merge_mesh(mesh1, mesh2, transfer_labeled_sets=True):
+    """merge two meshes (mesh.Mesh2D objects)"""
+    combined_coords = mesh1.coords.tolist()
+    # mapping for adjusting coord indices
+    mapping = {i: i for i in range(mesh2.num_nodes)}
+    # adjust connection indices for the second mesh using the mapping
+    def map_conn(conn, mapping):
+        return [mapping[idx] for idx in conn]
+
+    for idx, coord in enumerate(mesh2.coords):
+        # check if the point is close enough to any point in combined_coords
+        found = False
+        for i, existing_coord in enumerate(combined_coords):
+            if close_enough(coord[:2], existing_coord[:2], tolerance=1e-3):
+                mapping[idx] = i
+                found = True
+                break  
+        # if the point is not close enough to any existing point, add to combined_coords
+        if not found:
+            combined_coords.append(coord.tolist())
+            mapping[idx] = len(combined_coords) - 1
+
+    adjusted_conn2 = [map_conn(conn, mapping) for conn in mesh2.conn]
+
+    # merge conn lists
+    combined_conn = mesh1.conn + adjusted_conn2 
+
+    # merged mesh creation
+    m2_combined = watershed_workflow.mesh.Mesh2D(coords=np.array(combined_coords), conn=combined_conn)
+
+    # trasferring labeled sets
+    if transfer_labeled_sets:
+        m2_combined.labeled_sets = m2_combined.labeled_sets + mesh1.labeled_sets
```

Comment:
I think this breaks if m1 and m2 both have labeled sets that share the same label.  I'm not sure what to do about this -- probably there needs to be an option.  One could imagine two meshes, both have had land cover labeled sets based on NLCD. -- in that case, the merged mesh with LS = 51 should be the union of the two labeled sets.  Alternatively, one could imagine that the user just used a counter for soil types as the label, and therefore just wants to append the two labeled set lists and make the labels unique.

The key thing is that the merged list of labeled sets must have unique labels.

Reply:
hmm, it is tricky. We have to internally identify or put a flag if it is NLCD or soil, or broadly speaking, needs to be unionized or extended? Will that go in labeled_set class?

Reply:
I think the best bet is to compare names first.  If names are the same, then merge them.  Then compare labels.  If the labels are the same and the names are different, then the user probably doesn't care about the label and we can "relabel" one of them.

The reason being that the user will likely also be scripting input spec e.g. soil types, etc, that rely on the name.  And most of our names include some unique identifier (MUKEY, NLCD name, "surface", etc) that really is a classifier and we want them to be merged.

Reply:
Transferring labeledsets option will be added in the future. For now, meshes-to-be-merged should not have labeled-sets. They should be added after merging.